atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RowColumnRangeExtractor.java

```diff
@@ -68,57 +73,52 @@ public Map<byte[], Integer> getRowsToRawColumnCount() {
+            long startTs,
+            MetricsManager metricsManager) {
+        IdentityHashMap<byte[], LinkedHashMap<Cell, Value>> collector = new IdentityHashMap<>(canonicalRows.size());
+        IdentityHashMap<byte[], Column> rowsToLastCompositeColumns = new IdentityHashMap<>(canonicalRows.size());
+        IdentityHashMap<byte[], Integer> rowsToRawColumnCount = new IdentityHashMap<>(canonicalRows.size());
+        Set<byte[]> emptyRows = Collections.newSetFromMap(new IdentityHashMap<>(0));
 
-    RowColumnRangeExtractor(MetricsManager metricsManager) {
-        notLatestVisibleValueCellFilterCounter = metricsManager.registerOrGetCounter(
-                RowColumnRangeExtractor.class, AtlasDbMetricNames.CellFilterMetrics.NOT_LATEST_VISIBLE_VALUE);
-    }
+        // lazily create counter to avoid overhead when not needed
+        Supplier<Counter> notLatestVisibleValueCellFilterCounter =
+                Suppliers.memoize(() -> metricsManager.registerOrGetCounter(
+                        RowColumnRangeExtractor.class, CellFilterMetrics.NOT_LATEST_VISIBLE_VALUE));
 
-    public void extractResults(
-            Iterable<byte[]> canonicalRows, Map<ByteBuffer, List<ColumnOrSuperColumn>> colsByKey, long startTs) {
         // Make sure returned maps are keyed by the given rows
         Map<ByteBuffer, byte[]> canonicalRowsByHash = Maps.uniqueIndex(canonicalRows, ByteBuffer::wrap);
         for (Map.Entry<ByteBuffer, List<ColumnOrSuperColumn>> colEntry : colsByKey.entrySet()) {
             byte[] rawRow = CassandraKeyValueServices.getBytesFromByteBuffer(colEntry.getKey());
             byte[] row = canonicalRowsByHash.get(ByteBuffer.wrap(rawRow));
             List<ColumnOrSuperColumn> columns = colEntry.getValue();
 
-            if (!columns.isEmpty()) {
+            if (columns.isEmpty()) {
+                emptyRows.add(row);
+            } else {
                 rowsToLastCompositeColumns.put(
                         row, columns.get(columns.size() - 1).getColumn());
-            } else {
-                emptyRows.add(row);
             }
             rowsToRawColumnCount.put(row, columns.size());
             for (ColumnOrSuperColumn c : columns) {
                 Pair<byte[], Long> pair = CassandraKeyValueServices.decomposeName(c.getColumn());
-                internalExtractResult(startTs, row, pair.lhSide, c.getColumn().getValue(), pair.rhSide);
+                if (pair.rhSide < startTs) {
+                    Value previous = collector
+                            .computeIfAbsent(row, _bytes -> new LinkedHashMap<>()) // explicitly maintain returned order
+                            .put(
+                                    Cell.create(row, pair.lhSide),
+                                    Value.create(c.getColumn().getValue(), pair.rhSide));
+                    if (previous != null) {
+                        notLatestVisibleValueCellFilterCounter.get().inc();
```

Comment:
I _think_ the semantics are different to the past. The main one being that we put regardless, and increment the metric if the value existed beforehand. The difference that I am concerned about is that, in the past, if it already existed, we didn't overwrite the value - in this case, we do, blindly. I don't know this codepath very well, but I don't think that's what we want here.